### ISSUE TITLE
multiple parralel tasks for dp.la_en_all

### ISSUE THREAD

**Issue by kelson42** [openzim/zimfarm#813]
Many things are not understandable to me:
* We have a task of it running since a month https://farm.openzim.org/pipeline/60faed9f-264a-490c-9919-e33b96637623, so we have a Zimit bug here?
* when inlook to the recipe https://farm.openzim.org/recipes/dp.la_en_all, it seems we have 3 times the same scrape running in parallel, but this has not been scheduled manually?!
* I see we had 3 times success within the same hours. But the schedule is monthly?!

**Comment by kelson42** [comment]
Looks like to me there is something seriously wrong here? Hopefulky only wrong data in the DB.

**Comment by benoit74** [comment]
I'm working on it. Looks like there is a bug in the periodic scheduler.

**Comment by benoit74** [comment]
Impact seems fairly small, looks like there is no other occurence in the DB (I do not yet have the perfect SQL query to diagnose this easily)

**Comment by benoit74** [comment]
I built this query which looks for a `schedule` with more than one `task` with `updated_at` on the same day. It is not perfect because it does not take into account the situation where one task is `updated_at` few minutes before midnight and the next one is `updated_at` few minutes after.

```
SELECT schedule.name AS schedule_name,
    task.id,
    task.status,
    task.updated_at,
    task.requested_by,
    task.canceled_by
FROM task
    LEFT JOIN schedule ON task.schedule_id = schedule.id
WHERE schedule_id IN (
        SELECT schedule_id
        FROM (
                SELECT schedule_id,
                    date_trunc('day', task.updated_at) date_day,
                    COUNT(*) cnt
                FROM task
                WHERE status != 'failed'
                    and canceled_by IS NULL
                    and requested_by = 'period-scheduler'
                GROUP BY schedule_id,
                    date_day
            ) sub
        WHERE cnt > 1
    )
ORDER BY schedule_name,
    updated_at DESC
```

This returns only tasks 

```
schedule_name  id                                    status                   updated_at          requested_by      canceled_by
-------------  ------------------------------------  ---------------  --------------------------  ----------------  -----------
dp.la_en_all   2642a0ad-3a15-4e08-8e43-d9a2d6a43d32  succeeded        2023-08-28 02:56:29.418545  period-scheduler  <null>     
dp.la_en_all   653c7019-e5b9-407c-9107-0d3bb641be68  succeeded        2023-08-27 22:59:07.006753  period-scheduler  <null>     
dp.la_en_all   eaff89fa-0891-438c-b5cd-cac32a847845  succeeded        2023-08-27 22:48:08.264218  period-scheduler  <null>     
dp.la_en_all   06dd0c9e-d06d-474b-beda-36cc99281cd7  succeeded        2023-08-27 22:36:22.703242  period-scheduler  <null>     
dp.la_en_all   a6d210c8-6d3b-4548-bd48-7954e2747253  scraper_started  2023-08-27 22:00:57.22736   period-scheduler  <null>     
dp.la_en_all   ed86fe3d-ad79-4e64-9a56-0d8769df2803  failed           2023-07-28 13:30:02.735174  period-scheduler  <null>     
dp.la_en_all   9da1712e-1ce8-41d8-8724-ff393f6b47de  failed           2023-07-28 13:24:33.950324  period-scheduler  <null>     
dp.la_en_all   60faed9f-264a-490c-9919-e33b96637623  scraper_started  2023-07-27 18:01:12.771605  period-scheduler  <null>     
dp.la_en_all   c6e6ea97-0702-46ec-8071-e9c61cad5afd  failed           2023-07-27 17:27:17.367754  period-scheduler  <null>     
dp.la_en_all   5eca4e89-10c2-4140-8bb2-ec37b7bb5c58  failed           2023-07-27 17:20:22.817457  period-scheduler  <null>     
```

We can do the same for task with `updated_at` on the same `year-week` : 
```
SELECT schedule.name AS schedule_name,
    task.id,
    task.status,
    task.updated_at,
    task.requested_by,
    task.canceled_by
FROM task
    LEFT JOIN schedule ON task.schedule_id = schedule.id
WHERE schedule_id IN (
        SELECT schedule_id
        FROM (
                SELECT schedule_id,
                    CONCAT(date_trunc('year', task.updated_at), '-', date_trunc('week', task.updated_at)) date_week,
                    COUNT(*) cnt
                FROM task
                WHERE status != 'failed'
                    and canceled_by IS NULL
                    and requested_by = 'period-scheduler'
                GROUP BY schedule_id,
                    date_week
            ) sub
        WHERE cnt > 1
    )
ORDER BY schedule_name,
    updated_at DESC
```

This leads to the same result, i.e. no other situation.

So the problem is either linked to a recent change or due to a rare condition (either linked to schedule configuration or time-based/context-based conditions).
Last Zimfarm backend deployment is from Wed, 23 Aug 2023 14:32:03 +0200
Changes since 1st July are:
- small UI adaptations linked to Firefox on iPad bug
- rework on secret masking in API responses
- addition of maintenance scripts in Docker image
- addition of long_description field to Kolibri scraper
- update of zimit allowed devices list

So clearly it looks like a rare condition. What looks a bit weird is than the `most_recent_task_id` on the `schedule` is not correct.

**Comment by benoit74** [comment]
Let's extract all tasks from the culprit schedule `dp.la_en_all` (id is `c545d229-3b32-491c-98da-4f4d8efc204c`)

```
SELECT task_id,
    jsonb_extract_path_text(oneevent, 'code') as event_code,
    to_timestamp(
        cast(
            jsonb_extract_path_text(oneevent, 'timestamp', '$date') as numeric
        ) / 1000.0
    )::timestamp with time zone at time zone 'UTC' as event_date
FROM (
        SELECT id AS task_id,
            jsonb_array_elements(task.events) as oneevent
        FROM task
        WHERE schedule_id = 'c545d229-3b32-491c-98da-4f4d8efc204c'
    ) sub
ORDER BY event_date DESC
```


```
task_id                               event_code               event_date       
------------------------------------  -----------------  -----------------------
2642a0ad-3a15-4e08-8e43-d9a2d6a43d32  succeeded          2023-08-28 02:56:29.418
2642a0ad-3a15-4e08-8e43-d9a2d6a43d32  scraper_completed  2023-08-28 02:42:20.988
2642a0ad-3a15-4e08-8e43-d9a2d6a43d32  scraper_started    2023-08-28 00:02:10.416
2642a0ad-3a15-4e08-8e43-d9a2d6a43d32  started            2023-08-28 00:02:09.07 
2642a0ad-3a15-4e08-8e43-d9a2d6a43d32  reserved           2023-08-28 00:02:05.274
2642a0ad-3a15-4e08-8e43-d9a2d6a43d32  requested          2023-08-28 00:00:04.049
653c7019-e5b9-407c-9107-0d3bb641be68  succeeded          2023-08-27 22:59:07.006
eaff89fa-0891-438c-b5cd-cac32a847845  succeeded          2023-08-27 22:48:08.264
653c7019-e5b9-407c-9107-0d3bb641be68  scraper_completed  2023-08-27 22:47:08.206
eaff89fa-0891-438c-b5cd-cac32a847845  scraper_completed  2023-08-27 22:39:31.842
06dd0c9e-d06d-474b-beda-36cc99281cd7  succeeded          2023-08-27 22:36:22.703
06dd0c9e-d06d-474b-beda-36cc99281cd7  scraper_completed  2023-08-27 22:31:16.811
a6d210c8-6d3b-4548-bd48-7954e2747253  scraper_started    2023-08-27 22:00:57.227
a6d210c8-6d3b-4548-bd48-7954e2747253  started             2023-08-27 22:00:55.7 
a6d210c8-6d3b-4548-bd48-7954e2747253  reserved           2023-08-27 22:00:52.928
a6d210c8-6d3b-4548-bd48-7954e2747253  requested          2023-08-27 22:00:04.061
06dd0c9e-d06d-474b-beda-36cc99281cd7  scraper_started    2023-08-27 21:02:29.814
06dd0c9e-d06d-474b-beda-36cc99281cd7  started            2023-08-27 21:02:28.413
06dd0c9e-d06d-474b-beda-36cc99281cd7  reserved           2023-08-27 21:02:24.605
06dd0c9e-d06d-474b-beda-36cc99281cd7  requested          2023-08-27 21:00:03.982
eaff89fa-0891-438c-b5cd-cac32a847845  scraper_started    2023-08-27 20:02:33.427
eaff89fa-0891-438c-b5cd-cac32a847845  started            2023-08-27 20:02:31.978
eaff89fa-0891-438c-b5cd-cac32a847845  reserved           2023-08-27 20:02:28.106
eaff89fa-0891-438c-b5cd-cac32a847845  requested          2023-08-27 20:00:04.939
653c7019-e5b9-407c-9107-0d3bb641be68  scraper_started    2023-08-27 19:02:50.693
653c7019-e5b9-407c-9107-0d3bb641be68  started            2023-08-27 19:02:49.407
653c7019-e5b9-407c-9107-0d3bb641be68  reserved           2023-08-27 19:02:45.819
653c7019-e5b9-407c-9107-0d3bb641be68  requested          2023-08-27 19:00:03.952
ed86fe3d-ad79-4e64-9a56-0d8769df2803  failed             2023-07-28 13:30:02.735
9da1712e-1ce8-41d8-8724-ff393f6b47de  failed             2023-07-28 13:24:33.95 
9da1712e-1ce8-41d8-8724-ff393f6b47de  scraper_started    2023-07-28 04:03:15.043
9da1712e-1ce8-41d8-8724-ff393f6b47de  started            2023-07-28 04:03:10.129
9da1712e-1ce8-41d8-8724-ff393f6b47de  reserved           2023-07-28 04:03:02.283
9da1712e-1ce8-41d8-8724-ff393f6b47de  requested          2023-07-28 04:00:08.852
ed86fe3d-ad79-4e64-9a56-0d8769df2803  scraper_started    2023-07-28 03:00:55.902
ed86fe3d-ad79-4e64-9a56-0d8769df2803  started            2023-07-28 03:00:53.753
ed86fe3d-ad79-4e64-9a56-0d8769df2803  reserved           2023-07-28 03:00:48.844
ed86fe3d-ad79-4e64-9a56-0d8769df2803  requested          2023-07-28 03:00:08.317
60faed9f-264a-490c-9919-e33b96637623  scraper_started    2023-07-27 18:01:12.771
60faed9f-264a-490c-9919-e33b96637623  started            2023-07-27 18:01:11.319
60faed9f-264a-490c-9919-e33b96637623  reserved           2023-07-27 18:01:08.539
60faed9f-264a-490c-9919-e33b96637623  requested          2023-07-27 18:00:07.93 
c6e6ea97-0702-46ec-8071-e9c61cad5afd  failed             2023-07-27 17:27:17.367
5eca4e89-10c2-4140-8bb2-ec37b7bb5c58  failed             2023-07-27 17:20:22.817
c6e6ea97-0702-46ec-8071-e9c61cad5afd  scraper_started    2023-07-27 12:22:19.397
c6e6ea97-0702-46ec-8071-e9c61cad5afd  started            2023-07-27 12:22:15.238
c6e6ea97-0702-46ec-8071-e9c61cad5afd  reserved           2023-07-27 12:22:08.73 
c6e6ea97-0702-46ec-8071-e9c61cad5afd  requested          2023-07-27 12:00:08.329
5eca4e89-10c2-4140-8bb2-ec37b7bb5c58  scraper_started    2023-07-27 09:01:20.011
5eca4e89-10c2-4140-8bb2-ec37b7bb5c58  started            2023-07-27 09:01:17.661
5eca4e89-10c2-4140-8bb2-ec37b7bb5c58  reserved           2023-07-27 09:01:12.218
5eca4e89-10c2-4140-8bb2-ec37b7bb5c58  requested          2023-07-27 09:00:08.131
```

We observe that the situation already happened last month on 27-28th of July, and occured again yesterday / today 27-28th of August.

Unfortunately, backend logs are mostly already gone for the period of interest this morning. 

There are no more logs for the period where period task scheduler decided to start many times the same schedule.

We definitely need to implement something to keep logs at least for 7 days, this situation is made really worse to investigate due to this.

**Comment by benoit74** [comment]
OK, so I probably pin-pointed the issue.

There are two "guost" tasks which are still running on athena18 :
- https://farm.openzim.org/pipeline/60faed9f-264a-490c-9919-e33b96637623 (started in July)
- https://farm.openzim.org/pipeline/a6d210c8-6d3b-4548-bd48-7954e2747253 (started in August)

Those tasks are still pinging back every minute.
```
2023/08/28 05:12:38 [warn] 15#15: *176637 a client request body is buffered to a temporary file /var/cache/nginx/client_temp/0000041703, client: 100.64.6.28, server: , request: "PATCH /v1/tasks/60faed9f-264a-490c-9919-e33b96637623 HTTP/1.1", host: "api.farm.openzim.org"
INFO:common.utils:Task Container ping: 60faed9f-264a-490c-9919-e33b96637623
[pid: 24|app: 0|req: 27062/161583] 100.64.6.28 () {56 vars in 1351 bytes} [Mon Aug 28 05:12:38 2023] PATCH /v1/tasks/60faed9f-264a-490c-9919-e33b96637623 => generated 0 bytes in 152 msecs (HTTP/1.1 204) 2 headers in 99 bytes (7 switches on core 0)
100.64.6.28 - - [28/Aug/2023:05:12:39 +0000] "PATCH /v1/tasks/60faed9f-264a-490c-9919-e33b96637623 HTTP/1.1" 204 0 "-" "python-requests/2.31.0" "147.229.8.218"
2023/08/28 05:12:40 [warn] 15#15: *176502 a client request body is buffered to a temporary file /var/cache/nginx/client_temp/0000041704, client: 100.64.6.28, server: , request: "PATCH /v1/tasks/a6d210c8-6d3b-4548-bd48-7954e2747253 HTTP/1.1", host: "api.farm.openzim.org"
INFO:common.utils:Task Container ping: a6d210c8-6d3b-4548-bd48-7954e2747253
[pid: 26|app: 0|req: 73959/161584] 100.64.6.28 () {56 vars in 1351 bytes} [Mon Aug 28 05:12:40 2023] PATCH /v1/tasks/a6d210c8-6d3b-4548-bd48-7954e2747253 => generated 0 bytes in 153 msecs (HTTP/1.1 204) 2 headers in 99 bytes (8 switches on core 0)
100.64.6.28 - - [28/Aug/2023:05:12:41 +0000] "PATCH /v1/tasks/a6d210c8-6d3b-4548-bd48-7954e2747253 HTTP/1.1" 204 0 "-" "python-requests/2.31.0" "147.229.8.218"
```

Every time a task ping back the API, zimfarm updates the schedule "most_recent_task" and set it to the task which just ping back. So when more than one task is running at the same time, this is problematic. Every running task will become the `most_recent_task` in turn. This is particularly problematic when one of the task comes from very far in the past (more than the periodicity). If this old task is the `most_recent_task` when the periodic scheduler checks if a task has to be requested, then a new task is requested. And it can happens many times (once per hour) in a row. Then, it can suddenly stop to happen (like it is now) because the `most_recent_task` is really a recent one (just by luck of ping ordering, here we have only few seconds between ping and the second one is a recent one, but this is pure luck) and the periodic scheduler stops to request new tasks.

I proposes following enhancements:
1. change the periodic scheduler so that when it finds a schedule that has to be started because the `most_recent_task` is too old, it performs a check on the `most_recent_task` status. If the status is among the `incomplete` set, it does not request a new task and logs a warning instead
2. when the periodic scheduler finds such a situation, it sends a notification to Slack so that we investigate what has to be done
3. change the process of setting `most_recent_task`: instead of updating it every time an event occurs on a task, we could do it only a task creation (transformation from requested task to task)

Why the `zimit` scraper is stuck is still another point to investigate, I've opened a dedicated ticket to track this point: https://github.com/openzim/zimit/issues/210

@kelson42 @rgaudin WDYT?

**Comment by rgaudin** [comment]
Thanks for the thorough investigation.

The core of this issue is the fact that the scheduler started this task on a worker while it was already running on that worker. The rest are just collateral and I don't think we should focus on them now as it would only hide a deeper issue.

> change the periodic scheduler so that when it finds a schedule that has to be started because the most_recent_task is too old, it performs a check on the most_recent_task status. If the status is among the incomplete set, it does not request a new task and logs a warning instead

We've agreed in the past that a task running for 30d is an issue. We have no scraper that should exceed this (we used to) so maybe we should cancel such tasks in the periodic-tasks

> change the process of setting most_recent_task: instead of updating it every time an event occurs on a task, we could do it only a task creation (transformation from requested task to task)

I agree. I can't think of any reason not to.

Indeed without the zimfarm app log (I believe scheduler logs its decision –or it should), it's hard to know why it happened… but a month old is… old 😅

**Comment by benoit74** [comment]
I don't get why you consider that we are hiding a deeper issue, with my proposition, the log warning + the slack notification is exactly meant to avoid hiding the underlying issue. 

I consider this modification makes sense as a "guardrail", because I prefer to avoid bad things happening rather than expecting that all pieces fit together perfectly, especially because as pieces are modified, we can't expect every one of them to continue to fit perfectly together. But this is obviously personal taste.

I don't mind to automatically cancel tasks running for 30d, but we need a Slack notification as well on this action, otherwise the issue (a scraper running for 30d) is hidden. I nevertheless find it a bit risky. Are we sure we can automate such a thing, even for big scrapers (e.g. sotoki) which are running only few times a year? Maybe a slack notification which triggers a manual investigation would be enough.

**Comment by benoit74** [comment]
I'm sorry to read that my previous comment read as a lack of trust in zimfarm code. And I'm even more sorry if this offended you. To be transparent, I'm a bit offended about the fact that you consider I might not be comfortable with this issue and need you to implement it. My discussion is only about suggesting what I consider is the right thing to do, and trying to get a better understanding of why I do not get this is a bad thing to do. If we do not spend time on this kind of discussion, I'm left without room for improvement and I will have to always rely on very specific explanations from you on what has to be done, and probably also very lengthy code reviews with lot of things to fix.

In the current discussion, my comment is mostly influenced by a very long (and painful sometime) experience of the human (and beyond) inability to build a 100% perfect system, and the conviction that a system is resilient only if countermeasures are implemented to recover from bad things. 

The reasons why we got 2 tasks running and why it leads to a mix-up in `most_recent_task` are explained above, at least I consider I've understood what happened. As it looks like it is not really clear enough, I will reproduce it locally in order to provide a better understanding / assurance that my explanations are right.

Regarding my proposition to implement slack notifications, I don't get why you consider them only as "debug tools". From my perspective, they are long term solutions to either:
- inform us of what bad happens on the system and how the system recovered automatically. This is the case when the system decides to not start again a new task on a schedule which already has a task running. I agree that if our code and data are 100% perfect, this should never happen, but current situation proves that it happens). So as such, I prefer to have the system detect these abnormal situations which are important to detect (we should never have two tasks running at the same time on a single schedule), take the right decision (do not start a new task) and notify us (so that nothing is hidden, we can take manual actions if needed and we do not suddenly realize that the system automatically took a lot of decisions we did not expected it to take) 
- inform us that the system detected a situation which requires a manual decision, because we are not yet confident to build an automatic recover or because it is impossible / makes no sense. From my point of view, this is the case about the cancellation of tasks running for more than 30 days, because I have the feeling that the 30 days value is very arbitrary and I suspect we indeed need something simpler or more sophisticated. I consider that implementing an automatic recover might bite us at some point for big / long running scraper where I imagine we would be very sad to discover that the system cancelled a task which was almost complete and took a lot of our resources.

From my experience, relying on notifications in such situations is handy because:
- they notify us quickly (they do not require someone to open a web UI every day to check nothing bad is happening)
- they do not require new datastructures in the DB to store what has been detected / needs a decision + API to retrieve this information + UI changes to display it

Their downside is that they might only postpone a decision which we do not know how to take today (e.g. should we cancel all tasks that have ran for 30+ days, which will still be a decision to make once the notification is received). To cover this downside my experience showed that deciding on a very clear and straightforward process (which could be done by any new-comer) for every notification we receive is very important / essential.

I agree that notifications alone are not enough, and in the current issue we have to fix the compute of the `most_recent_task`, and at least we are both ok on this.

Now that I've clearly (I hope) exposed the reasons why I'm suggesting to add notifications in both situations, I can accept that you decide they are not needed / useless / harmful but I would really like to understand the underlying "why".

**Comment by rgaudin** [comment]
I'm sorry about what I said. I felt like it required painful setup and investigation and I didn't want to impose that on you, especially as I had you tired of ZF bugs.

Thanks to your investigation, we know there's no emergency and we can even disable the recipe until we fix it all.

> we got 2 tasks running and why it leads to a mix-up in most_recent_task are explained above

I'm sorry but I don't see it. I read that both running tasks sets themselves as the most recent one. This is a bug and you've proposed a good solution for that. How come we've had two task from the same recipe on the same worker is what I'm trying to understand. If you found out, please share as I did not understand from the previous comments. If you don't, that's one important bug we should investigate as the scheduler is core to zimfarm obviously.

>  I don't get why you consider them only as "debug tools"

You want a _somewhat identified bug_ to be reported: it's identified enough that you know where to put your notification code but you want to know the state of the environment (values of other variables in fact). That's pretty much debugging to me.

Everything you describe in the first paragraph is a log system. We log into the void but we know we need a centralized platform (https://github.com/kiwix/k8s/issues/5). If we're at the point where the lack of logs is problematic (and it seems we are), then it might be a good opportunity to tackle this issue, which will benefit all projects. With such a tool in place, we'll be able to set alerts and all.

> inform us that the system detected a situation which requires a manual decision

I don't think the zimfarm is a system that requires or should require a manual decision (from developers!). As I said, this 30d delay, arbitrary indeed, has been discussed. It's not enforced at the moment but we know from experience that a task this long is a bug either on the worker or on the scraper. 30d is a very long time and the Content team should probably have raised it and killed it before you did. If they don't then the system could cancel it when due and set the appropriate reason. Whenever it's due to run again, it's attempted again. Nothing's hidden.

ZF developer should be informed about zimfarm bugs, not recipes/scraper issues (in this context 😅)

>  I would really like to understand the underlying "why".

To me, this is a scope and responsibility issue. We need to know what ZF does so ZF needs to tell. It does, in the logs. We don't have access to logs older than a ridiculous delay so it's clear that we need to fix this instead of implementing a parallel reporting system in the zimfarm. A lot of our other tools (cardshop, cms) share the same needs.

That's my opinion but it looks like more of a process question here so I'am expecting @kelson42 feedback.

**Comment by kelson42** [comment]
@rgaudin @benoit74 Because of lack of time I don't fully understand the situation IMO. Here a few remarks:
* Please make a call about the problems ariund this recipe
* It seems to me I agree with @rgaudin latest comment
* We should not re-implement a log manager (features)
* If a quick improvement around logging can be done to allow a diagnostic, maybe we should still consider it.

**Comment by benoit74** [comment]
Let me rephrase my understanding of the situation / supposition of the cause, looks like my comments have not been as clear as I expected (sorry). 

This detailed explanation is based on few suppositions because we miss some logs, so I will reproduce the issue locally to confirm this is something that could really happen, but I'm really confident (maybe too much 😢)

- at `2023-08-27 19:00:00`, the `periodic-scheduler` looks for schedules which needs to be requested again, like every hour
  - at that time, `dp.la_en_all` schedule has the task `60fae` which is still running, pinging back every minute so ZF consider it as the `most_recent_task`
  - the `periodic-scheduler` looks into this `most_recent_task` to find its `started` timestamp and checks if it is older than the schedule periodicity
  - since `60fae` was started at `2023-07-27 18:01:11`, it is older than the schedule periodicity so the `periodic-scheduler` requests a new task `653c7` at `2023-08-27 19:00:03`
  - few minutes later, task `653c7` starts and we now have two tasks running on the same schedule
  - every minute, both tasks ping back ZF to inform they are still alive and ZF makes them the `most_recent_task` in turn
- at `2023-08-27 20:00:00`, the `periodic-scheduler` looks again for schedules which needs to be requested again
  - the last task to ping back the ZF was `60fae` so it is considered as the `most_recent_task`, its start time is still too far in the past so the `periodic-scheduler` decides again to request schedule with task `eaff8`
  - few minutes later, task `eaff8` starts and we now have three tasks running on the same schedule
- at `2023-08-27 21:00:00`, the scenario repeats again and we now have task `06dd0`
- at `2023-08-27 22:00:00`, the scenario repeats again and we now have task `a6d21`
- at `2023-08-27 22:36:22`, task `06dd0`  is marked as `succeeded` 
- at `2023-08-27 22:48:08`, task `eaff8` is marked as `succeeded`
- at `2023-08-27 22:59:07`, task `653c7` is marked as `succeeded` 
- at `2023-08-27 23:00:00`, no new task is requested, this time the `most_recent_task` was probably one of the recent ones (`653c7`probably since it just succeeded few seconds ago)
- at `2023-08-28 00:00:00`, the scenario repeats again and we now have task `2642a`
- the scenario does not repeat anymore, probably due to timing
- at `2023-08-28 02:56:29`, task `2642a` is marked as `succeeded`
  - at this point, we have two tasks still active: `60fae` (from July) and `a6d21` (from August)
- at `2023-08-28 05:12`, I have logs that show that the `60fae` task pings back at xx:xx:38 seconds and `a6d21` pings back at xx:xx:40 seconds
  - so when the `periodic-scheduler` starts at xx:xx:00 seconds, the `most_recent_task` is `a6d21`  which is recent and no task is requested again
  - this has probably been the situation since `02:56`

Such a situation seems to already have occurred in July, but we do not have enough task history to confirm.

Since I cancelled both long running tasks and the upstream bug is fixed, the situation is now back to normal and the `periodic-scheduler` should start only on task in a bit less than one month, just like it does for other schedules.

This issue nevertheless proves that the situation is fragile because ZF does not have any safeguard preventing yesterday situation to occur again. Should a task run for more than 30 days on a monthly schedule, situation will occur again. It is nevertheless the first occurrence, so priority is moderate.

Another topic we might need to have a look into it the significant variability of task duration. Recent task history from yesterday show that task duration is significantly variable ; `a6d21` was still running after 11 hours while `06dd0` completed in less than 2 hours. I don't know if this needs an investigation or not. ❓ Should I open an issue on this topic❓ 

I will reproduce scenario above on my machine to confirm this is a good explanation of yesterday situation. 

@rgaudin is not working today, I'm not working tomorrow, but we can have a live chat on this Thursday, sure.

**Comment by kelson42** [comment]
My two cents:
* Forbidding the auto-scheduler to launch a task when still running LGTM
* Not been able to start a scrape twice manually sounds to be an inconvenient.... in particular because killing a task might take many hpurs under certain conditions. I would not forbid this.
* I'm not in favour of killing task arbitrarly after 30 days because seems very intrusive and not necessary.

**Comment by rgaudin** [comment]
Thank you for the explanation @benoit74; indeed the key to all this is that we don't check for the existence a previously running task when requesting. That's not done on purpose due to the 30d delay thing we discussed. It's just a missing check.

I share all of @kelson42's comments. Scheduler should not create a task if one is already running but manual triggers are OK. 30d task being a bug is kept as a rule of thumb but not enforced in any way.

**Comment by benoit74** [comment]
FYI the output of the live discussion with @rgaudin that we will implement soon:
- ZF must not schedule automatically a task if the `most_recent_task` of the schedule is still running
- ZF must update `most_recent_task` only when the task is requested (if feasible, maybe not because task and requested tasks are not same objects, and the fact that the `most_recent_task` could be either a task or a requested task might raise lots of issues) or when the task is created (i.e. transform from requested task to task because it has been assigned to a worker)
- Monitoring the duration of Zimfarm tasks is probably something that must be done by the content team because it is highly context sensitive (to be discussed at next weekly, outside the scope of this issue)
